### ISSUE TITLE
📢 IMPORTANT Advisory for Mac Users

### ISSUE THREAD

**Issue by damithc** [nus-cs2103-AY2122S1/forum#231]
This advisory applies to **all M1 Mac users (and those who used Rosetta to get their iP to work)**

It turns out that you might not be able to run the JAR files produced by some of the classmates due to various incompatibility issues between your environment and theirs, although you may have got your own JAR file to work in your computer. Some of our tutors looked into this problem and found a specific build of OpenJDK that doesn't seem to have these problems. If you are among the affected Mac users, please switch to that version of OpenJDK by following the instructions below. 

Note: While it's not necessary to do this switch _before_ the iP release, you should switch by week 7, or else you may not be able to run iP JAR files created by others (which is required for the iP peer evaluation that is coming up in week 8, and for the tP  practical exam later in the semester).

**Instructions:**
Download and install the Azul build of OpenJDK 11 found [here](https://www.azul.com/downloads/?version=java-11-lts&os=macos&architecture=arm-64-bit&package=jdk-fx) (scroll to the bottom of the page) -- it supports fat JAR files compiled by all other systems that we have tested.

![image](https://user-images.githubusercontent.com/1673303/133399726-93f98ee4-6efb-4f37-830d-46a72298ab49.png)

Recommended: Remove the other versions of OpenJDK so that you do not accidentally launch the JAR using the wrong version

After installation, the command `java --version` should give the following output:
```
openjdk 11.0.12 2021-07-20 LTS
OpenJDK Runtime Environment Zulu11.50+19-CA (build 11.0.12+7-LTS)
OpenJDK 64-Bit Server VM Zulu11.50+19-CA (build 11.0.12+7-LTS, mixed mode)
```

And `whereis java` should output `/usr/bin/java`

Recommended: Set the `JAVA_HOME`/`JDK_HOME` system variables to point to `/Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home`

If you encounter problems while doing the above or need more info, please reply to this post.

**Comment by sivayogasubramanian** [comment]
Hi Prof @damithc, may I know what exactly is the problem here? I have tested jar files for my friends for iP, one of them using windows and another MacOS catalina (intel system) and I have no problem running them under rosetta.

Edit: I realised that both of them are using custom CSS files to specify fonts that are available in all systems. For those who are facing this problem, you can follow [this video](https://www.youtube.com/watch?v=wwV_L3lKYYw) to uninstall java first before installing the version mentioned by prof @damithc  above. I tested a jar file that does not set the font, it seems to work.

**Comment by damithc** [comment]
> Edit: I realised that both of them are using custom CSS files to specify fonts that are available in all systems. For those who are facing this problem, you can follow [this video](https://www.youtube.com/watch?v=wwV_L3lKYYw) to uninstall java first before installing the version mentioned by prof @damithc above. I tested a jar file that does not set the font, it seems to work.

@sivayogasubramanian Yes, Rosetta might might work for some cases. Thanks for the additional info and thanks for confirming the given JDK version works.

**Comment by Xiaoyunnn** [comment]
Hi prof, I followed the steps in [this video](https://www.youtube.com/watch?v=wwV_L3lKYYw) to uninstall the existing jdk but when I run `java -version`, there is still java installed on my mac...
<img width="646" alt="Screenshot 2021-09-16 at 7 17 00 PM" src="https://user-images.githubusercontent.com/76084739/133602806-869ae431-a4d9-4494-a2d4-32134f3cb7b1.png">

**Comment by Zzhiwei** [comment]
@damithc  
I can't remove this version of java that is installed on my m1 mac
```
openjdk 16 2021-03-16
OpenJDK Runtime Environment (build 16+36-2231)
OpenJDK 64-Bit Server VM (build 16+36-2231, mixed mode, sharing)
```
I followed the video recommended by @sivayogasubramanian by deleting all current jdk in `/Library/Java/JavaVirtualMachines`  and then installed the Azul build but when i enter
```
java --version
```
The same version above still shows up.
Any idea how to resolve this?

Additional note:
The openjdk 16 is not even in the `/Library/Java/JavaVirtualMachines` directory

**Comment by damithc** [comment]
@softmagnet I'm not a Mac user myself. Let's wait to see if any Mac user has any suggestions.

**Comment by sivayogasubramanian** [comment]
@softmagnet Maybe you can try uninstalling with brew?

**Comment by nicole-luo-exe** [comment]
@softmagnet Did you do` cd Library/Java/JavaVirtualMachines` instead of  ` cd /Library/Java/JavaVirtualMachines`?  The first one refers to /Users/xxxx/Library/Java/JavaVirtualMachines which is different from /Library/Java/JavaVirtualMachines.

**Comment by mingyi456** [comment]
@softmagnet What does `whereis java` and `which java` give you?

**Comment by Zzhiwei** [comment]
> @softmagnet What does `whereis java` and `which java` give you?

it gives me `/usr/bin/java`

> @softmagnet Did you do` cd Library/Java/JavaVirtualMachines` instead of ` cd /Library/Java/JavaVirtualMachines`? The first one refers to /Users/xxxx/Library/Java/JavaVirtualMachines which is different from /Library/Java/JavaVirtualMachines.

ok i found all the diff versions of jdk, they are at ` cd /Library/Java/JavaVirtualMachines` instead of ` cd Library/Java/JavaVirtualMachines` . Deleting these versions have solved my issue. Much thanks!

**Comment by NicolasCwy** [comment]
@sivayogasubramanian What is rosetta? I am on an intel chip Mac, but the person I am reviewing has font issues. Would like to try to use rosetta to fix the font issue, as it seems that the java version mentioned above is not applicable to me

**Comment by sivayogasubramanian** [comment]
> @sivayogasubramanian What is rosetta? I am on an intel chip Mac, but the person I am reviewing has font issues. Would like to try to use rosetta to fix the font issue, as it seems that the java version mentioned above is not applicable to me

Rosetta 2 enables a Mac with Apple silicon (M1 chip) to use apps built for a Mac with an Intel processor. You should not need rosetta. You can try downloading the above JDK. But remember to change the architecture to x86 64 bit.

**Comment by NicolasCwy** [comment]
@sivayogasubramanian Thanks using the Azul build of OpenJDK 11 for x86 64 bit worked!

**Comment by kaixin-hc** [comment]
I'd also like to share this Stack Overflow post: https://stackoverflow.com/questions/21964709/how-to-set-or-change-the-default-java-jdk-version-on-macos which helped me figure out where my version of java was being stored. If you're too lazy to read, `/usr/libexec/java_home -V` will show where your java is located, and then you can remove the wrong java via `rm -rf <java version here>` (I didn't need sudo)

**Comment by s7u4rt99** [comment]
> I'd also like to share this Stack Overflow post: https://stackoverflow.com/questions/21964709/how-to-set-or-change-the-default-java-jdk-version-on-macos which helped me figure out where my version of java was being stored. If you're too lazy to read, `/usr/libexec/java_home -V` will show where your java is located, and then you can remove the wrong java via `rm -rf <java version here>` (I didn't need sudo)

Thanks for this, it helped me find my JDK. Mine happened to be in /Users/xxxx/Library/Java/JavaVirtualMachines and not /Library/Java/JavaVirtualMachines. Spent quite a while looking for it.

**Comment by s7u4rt99** [comment]
Hi guys, I followed the steps and am currently on the correct version of JDK. However, when I run the code to run the duke.jar file to evaluate the person assigned to me, the font issue is still not fixed. What do I do?

I am currently on the mac intel so I downloaded the x86 64 version.

<img width="552" alt="Screenshot 2021-10-06 at 9 21 12 PM" src="https://user-images.githubusercontent.com/77237808/136210496-d8e9b759-7673-4268-8e85-f4fa88a6fdb9.png">

@NicolasCwy May I know if you encountered the same issue as me and how you fixed it?
> @sivayogasubramanian Thanks using the Azul build of OpenJDK 11 for x86 64 bit worked!

**Comment by NicolasCwy** [comment]
> Hi guys, I followed the steps and am currently on the correct version of JDK. However, when I run the code to run the duke.jar file to evaluate the person assigned to me, the font issue is still not fixed. What do I do?
> 
> 
> 
> I am currently on the mac intel so I downloaded the x86 64 version.
> 
> 
> 
> <img width="552" alt="Screenshot 2021-10-06 at 9 21 12 PM" src="https://user-images.githubusercontent.com/77237808/136210496-d8e9b759-7673-4268-8e85-f4fa88a6fdb9.png">
> 
> 
> 
> @NicolasCwy May I know if you encountered the same issue as me and how you fixed it?
> 
> > @sivayogasubramanian Thanks using the Azul build of OpenJDK 11 for x86 64 bit worked!
> 
> 
> 
> 

@s7u4rt99   Did you set the environment variables JAVA_HOME and JDK_HOME to `/Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home`. You can follow this [tutorial](https://www.serverlab.ca/tutorials/linux/administration-linux/how-to-set-environment-variables-in-linux/) if you have not. Hope this helps!

**Comment by s7u4rt99** [comment]
> @s7u4rt99 Did you set the environment variables JAVA_HOME and JDK_HOME to `/Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home`. You can follow this [tutorial](https://www.serverlab.ca/tutorials/linux/administration-linux/how-to-set-environment-variables-in-linux/) if you have not. Hope this helps!

Yes i did `export JAVA_HOME=/Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home` and `export JDK_HOME=/Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home` already but still the fonts are not fixed.

**Comment by NicolasCwy** [comment]
> > @s7u4rt99 Did you set the environment variables JAVA_HOME and JDK_HOME to `/Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home`. You can follow this [tutorial](https://www.serverlab.ca/tutorials/linux/administration-linux/how-to-set-environment-variables-in-linux/) if you have not. Hope this helps!
> 
> Yes i did `export JAVA_HOME=/Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home` and `export JDK_HOME=/Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home` already but still the fonts are not fixed.

Does `whereis java` and `which java` give you ` /usr/bin/java` as mentioned above? Also sounds kinda silly, but maybe try restart your computer. Else I'm not too sure either sorry!

**Comment by s7u4rt99** [comment]
> Does `whereis java` and `which java` give you ` /usr/bin/java` as mentioned above? Also sounds kinda silly, but maybe try restart your computer. Else I'm not too sure either sorry!

Yea both gives me `/usr/bin/java` and I also restarted my computer already. No worries man! Thanks for helping. Hoping someone else or prof @damithc knows and can advise me what to do.

**Comment by mingyi456** [comment]
At this point, is it a bug with the jar file or the code itself? Maybe try to get some other people to send you their jar files for you to test.

**Comment by damithc** [comment]
> At this point, is it a bug with the jar file or the code itself? Maybe try to get some other people to send you their jar files for you to test.

This is a good idea. @s7u4rt99 Also, do you have the same problem with the 2nd JAR file?

**Comment by s7u4rt99** [comment]
> > At this point, is it a bug with the jar file or the code itself? Maybe try to get some other people to send you their jar files for you to test.
> 
> This is a good idea. @s7u4rt99 Also, do you have the same problem with the 2nd JAR file?

Yes, I've tested with both JAR files and my friend's one but sadly the fonts are still unreadable.

**Comment by kouyk** [comment]
> Yes, I've tested with both JAR files and my friend's one but sadly the fonts are still unreadable.

Perhaps we need to identify a known good first. You might want to try running the latest version of AB3 and also one release before and observe what happens.

Also, I assume you're running it using `java -jar` right? In this case can you provide the CLI output as well? Logs are usually much more descriptive and helpful in diagnosing the problem.

**Comment by s7u4rt99** [comment]
> Perhaps we need to identify a known good first. You might want to try running the latest version of AB3 and also one release before and observe what happens.
> 
> Also, I assume you're running it using `java -jar` right? In this case can you provide the CLI output as well? Logs are usually much more descriptive and helpful in diagnosing the problem.

Yes, I am running it using `java -jar`. The latest AB3 works fine and these are the logs.

<img width="735" alt="Screenshot 2021-10-07 at 12 48 45 AM" src="https://user-images.githubusercontent.com/77237808/136248251-64c87b1a-ac4b-4926-9b6f-0ea47d403661.png">
<img width="1429" alt="Screenshot 2021-10-07 at 12 45 39 AM" src="https://user-images.githubusercontent.com/77237808/136247784-b4af05e4-398b-4525-b84b-f25a56190459.png">

The release before has the unreadable fonts.
<img width="735" alt="Screenshot 2021-10-07 at 12 47 06 AM" src="https://user-images.githubusercontent.com/77237808/136247993-1a504c1c-d932-463c-a14f-18242a9c436f.png">

And this is the log.

<img width="1438" alt="Screenshot 2021-10-07 at 12 48 00 AM" src="https://user-images.githubusercontent.com/77237808/136248123-cbe6cad1-8ba5-409e-b03d-6756eb3be962.png">

**Comment by kouyk** [comment]
Can I confirm the version of MacOS that you're running? And can I also confirm if you installed the JDK FX version or JDK version from Azul?

AFAIK, this particular bug has been fixed in 2019 and shouldn't be a problem anymore after JavaFX 11.0.8, see https://bugs.openjdk.java.net/browse/JDK-8234916

> And this is the log.

BTW it is recommended that you paste the text output of the log if it is short, or attach as a text file. It is considered a bad practice to simply post screenshots of CLI output especially since it makes it difficult to search for.

**Comment by s7u4rt99** [comment]
> BTW it is recommended that you paste the text output of the log if it is short, or attach as a text file. It is considered a bad practice to simply post screenshots of CLI output especially since it makes it difficult to search for.

Sorry for the inconvenience caused by screenshotting it instead of copying it over, it slipped my mind. Will copy and paste it next time.

> Can I confirm the version of MacOS that you're running? And can I also confirm if you installed the JDK FX version or JDK version from Azul?

I think this is it. I went to download the JDK FX version again and it is working now. I must have accidentally downloaded the JDK one earlier....twice.... Thank you for helping me catch this @kouyk ! If not I would not have solved it so quickly.

**Comment by damithc** [comment]
> I think this is it. I went to download the JDK FX version again and it is working now. I must have accidentally downloaded the JDK one earlier....twice.... 

Thanks everyone who pitched in @kouyk @mingyi456 @NicolasCwy

**Comment by willyamped** [comment]
Hi all. I know this is quite late, but does this method also fix font issues when the program is run from IntelliJ?

If I open a .jar file, there is no font issues. However, I still encounter the font issues from IntelliJ.

I have followed all the necessary steps, including downloading the correct OpenJDK version (11.0.12+7, JDK FX, I tried both ARM 64-bit and x86 64-bit), setting the `JAVA_HOME` and `JDK_HOME` paths. When I write `whereis java` and `which java`. It also gives me the correct output. Thank you. Really appreciate the help!

**Comment by bharathcs** [comment]
Hey @willyamped I think what's happening there is that IntelliJ is calling up the wrong java version. You should check Project SDK under Project Structure with your tp folder opened in IntelliJ.

**Comment by willyamped** [comment]
@bharathcs Thank you for the prompt reply! You are right! Thank you so much. I only changed the OpenJDK version not the Project SDK version :'). Thank you! 👍
